cudax/include/cuda/experimental/__container/vector.cuh

```diff
@@ -0,0 +1,1585 @@
+    {
+      this->__copy_same(__other.__unwrapped_begin(), __other.__unwrapped_end(), __unwrapped_begin());
+    }
+    else
+    { // We need to reallocate and copy. Note we do not change the size of the current vector until the copy is done
+      __buffer __new_buf{__other.get_resource(), __other.__size_};
+      this->__copy_same(__other.__unwrapped_begin(), __other.__unwrapped_end(), __new_buf.begin());
+      __buf_.__swap_allocations(__new_buf);
+    }
+
+    __size_ = __other.__size_;
+    return *this;
+  }
+
+  //! @brief Move-assigns a vector
+  //! @param __other The other vector.
+  //! Clears the vector and swaps the contents with \p __other.
+  vector& operator=(vector&& __other) noexcept
+  {
+    if (this == _CUDA_VSTD::addressof(__other))
+    {
+      return *this;
+    }
+
+    __buf_  = _CUDA_VSTD::move(__other.__buf_);
+    __size_ = _CUDA_VSTD::exchange(__other.__size_, 0);
+    return *this;
+  }
+
+  //! @brief Assigns an initializer_list to a vector, replacing its content with that of the initializer_list
+  //! @param __ilist The initializer_list to be assigned
+  vector& operator=(_CUDA_VSTD::initializer_list<_Tp> __ilist)
+  {
+    const auto __count = __ilist.size();
+    if (capacity() < __count)
+    {
+      __buffer __new_buf{get_resource(), __count};
+      this->__copy_cross(__ilist.begin(), __ilist.end(), __new_buf.begin(), __count);
+      __buf_.__swap_allocations(__new_buf);
+    }
+    else
+    {
+      this->__copy_cross(__ilist.begin(), __ilist.end(), __unwrapped_begin(), __count);
+    }
+    __size_ = __count;
+    return *this;
+  }
+
+  //! @brief Copy-assigns from a different vector
+  //! @param __other The other vector.
+  _LIBCUDACXX_TEMPLATE(class... _OtherProperties)
+  _LIBCUDACXX_REQUIRES(__properties_match<_OtherProperties...>)
+  vector& operator=(const vector<_Tp, _OtherProperties...>& __other)
+  {
+    if (this == reinterpret_cast<const vector*>(_CUDA_VSTD::addressof(__other)))
+    {
+      return *this;
+    }
+
+    const auto __count = __other.size();
+    if (capacity() < __count)
+    {
+      __buffer __new_buf{get_resource(), __count};
+      this->__copy_cross<const_pointer, __transfer_kind<_OtherProperties...>>(
+        __other.__unwrapped_begin(), __other.__unwrapped_end(), __new_buf.begin(), __count);
+      __buf_.__swap_allocations(__new_buf);
+    }
+    else
+    {
+      this->__copy_cross<const_pointer, __transfer_kind<_OtherProperties...>>(
+        __other.__unwrapped_begin(), __other.__unwrapped_end(), __unwrapped_begin(), __count);
+    }
+    __size_ = __count;
+    return *this;
+  }
+
+  //! @brief Move-assigns from a different vector
+  //! @param __other The other vector.
+  _LIBCUDACXX_TEMPLATE(class... _OtherProperties)
+  _LIBCUDACXX_REQUIRES(__properties_match<_OtherProperties...>)
+  vector& operator=(vector<_Tp, _OtherProperties...>&& __other)
+  {
+    if (this == reinterpret_cast<vector*>(_CUDA_VSTD::addressof(__other)))
```

Comment:
when would two vectors with different types have the same address??

Reply:
I had the hope that we can just downcast a vector with a set of properties to any vector with a subset